### PR TITLE
Task-25-Linkedin-Auth-Front-End-/-Back-End

### PR DESCRIPTION
**Pull request by melonholy**

### CODE REVIEW

File: api/controllers/auth.js [hunk 1]
```diff
@@ -11,8 +11,13 @@ const login = async (req, res) => {
 };
 
 const googleLogin = async (req, res) => {
-  const response = await services.googleLogin(req.body);
-  res.send(response);
+  const response = await services.googleLogin(req.user);
+  res.redirect('http://localhost:3000/?tokenGoogle=' + response);
```
Comment (logoped583st) [inline api/controllers/auth.js]: hardcode

File: api/controllers/auth.js [hunk 2]
```diff
@@ -11,8 +11,13 @@ const login = async (req, res) => {
...
+};
+
+const linkeinLogin = async (req, res) => {
+  const response = await services.linkeinLogin(req.user);
+  res.redirect('http://localhost:3000/?tokenLinkedin=' + response);
```
Comment (logoped583st) [inline api/controllers/auth.js]: hardcode

File: api/helpers/jwtHelpers.js [hunk 3]
```diff
@@ -4,9 +4,13 @@ const generateToken = ({ id, email }) => {
   const token = jwt.sign({ id, email }, process.env.SECRET, {
     expiresIn: process.env.EXPIRES_IN,
   });
-  return { token };
+  return token;
 };
 
+const decodeToken = token => {
```
Comment (logoped583st) [inline api/helpers/jwtHelpers.js]: not used

File: passport.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  });
+  passport.deserializeUser((user, done) => {
+    done(null, user);
+  });
+  passport.use(
+    new GoogleStrategy(
+      {
+        clientID: process.env.GOOGLE_CLIENT_ID,
+        clientSecret: process.env.GOOGLE_SECRET,
+        callbackURL: 'http://localhost:8080/api/users/auth/google/callback',
```
Comment (logoped583st) [inline passport.js]: hardcode

File: api/controllers/auth.js (hunk 1 above)
Comment (OlegPanasyuk) [reply api/controllers/auth.js]: Use .env variables in this case `const HOST = process.env.HOST || 'http://locahost:3000/';`

File: api/helpers/jwtHelpers.js [hunk 5]
```diff
@@ -4,9 +4,13 @@ const generateToken = ({ id, email }) => {
   const token = jwt.sign({ id, email }, process.env.SECRET, {
     expiresIn: process.env.EXPIRES_IN,
   });
-  return { token };
+  return token;
```
Comment (OlegPanasyuk) [inline api/helpers/jwtHelpers.js:7]: Minify code, please 
```
return jwt.sign({ id, email }, process.env.SECRET, {
    expiresIn: process.env.EXPIRES_IN,	    expiresIn: process.env.EXPIRES_IN,
});	  });
```

File: api/routes/auth.js [hunk 6]
```diff
@@ -1,9 +1,32 @@
 const express = require('express');
 const router = express.Router();
 const controllers = require('../controllers/auth');
+const passport = require('passport');
 
 router.post('/login', controllers.login);
 
 router.post('/registration', controllers.registration);
-router.post('/google/auth', controllers.googleLogin);
+router.get('/linkedin/auth', (req, res) => {
```
Comment (OlegPanasyuk) [inline api/routes/auth.js]: Push middlewares to file in middleware's folder. Controllers in one's folder.

File: api/services/auth.js [hunk 7]
```diff
@@ -12,23 +12,22 @@ const login = async ({ email, password }) => {
...
-    email: data.payload.profileObj.email,
-    token: data.payload.Zi.id_token,
-  };
-  registration(payload);
-  return jwtHelpers.generateToken(payload);
+  registration(data);
+  return jwtHelpers.generateToken(data);
 };
+
+const linkeinLogin = async data => {
```
Comment (OlegPanasyuk) [inline api/services/auth.js]: Destruct necessary data by the fields.
`const {email, password} = data`

File: api/services/auth.js [hunk 8]
```diff
@@ -12,23 +12,22 @@ const login = async ({ email, password }) => {
...
+  registration(data);
```
Comment (OlegPanasyuk) [inline api/services/auth.js:16]: What happened if ` registration` failed? 
If `generateToken` executed earlier then `registration`?

File: index.js [hunk 9]
```diff
@@ -1,7 +1,11 @@
 require('dotenv').config();
-const cors = require('cors')
+
+const cors = require('cors');
 const express = require('express');
 const bodyParser = require('body-parser');
+const passport = require('passport');
+
+const session = require('express-session')
```
Comment (OlegPanasyuk) [inline index.js]: Are we use sessions?

File: passport.js [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        done(null, userData);
+      }
+    )
+  );
+  passport.use(
+    new LinkedInStrategy(
+      {
+        clientID: process.env.LINKEDIN_CLIENT_ID,
+        clientSecret: process.env.LINKEDIN_SECRET,
+        callbackURL: 'http://localhost:8080/api/users/auth/linkedin/callback',
```
Comment (OlegPanasyuk) [inline passport.js]: Push this host to the variable.

File: passport.js [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        scope: ['r_liteprofile', 'r_emailaddress'],
+      },
+      (accessToken, refreshToken, profile, done) => {
+        const userData = {
+          id: profile.id,
```
Comment (OlegPanasyuk) [inline passport.js]: Use destructurisation

File: api/middlewares/refreshAccessToken.js [hunk 12]
```diff
@@ -2,15 +2,15 @@ const jwtDecode = require('jwt-decode');
...
-    let decodedToken = jwtDecode(req.headers.authorization);
-    const user = {
-      id: decodedToken.sub,
-      name: decodedToken.name,
-      expiresIn: process.env.EXPIRES_IN,
-    };
-    const token = jwt.sign(user, process.env.SECRET);
-    req.headers.authorization = token;
-    next();
+  let decodedToken = jwtDecode(req.headers.authorization);
```
Comment (slavasgod) [inline api/middlewares/refreshAccessToken.js]: Use const instead of let

File: api/middlewares/refreshAccessToken.js [hunk 13]
```diff
@@ -2,15 +2,15 @@ const jwtDecode = require('jwt-decode');
 const jwt = require('jsonwebtoken');
 
 const refreshAccessToken = (req, res, next) => {
```
Comment (slavasgod) [inline api/middlewares/refreshAccessToken.js:3]: Migrate this to middlewares

File: api/middlewares/refreshAccessToken.js [hunk 14]
```diff
@@ -2,15 +2,15 @@ const jwtDecode = require('jwt-decode');
...
+    name: decodedToken.name,
+    expiresIn: process.env.EXPIRES_IN,
   };
-  
-module.exports = refreshAccessToken;
\ No newline at end of file
+  const token = jwt.sign(user, process.env.SECRET);
+  req.headers.authorization = token;
+  next();
+};
```
Comment (slavasgod) [inline api/middlewares/refreshAccessToken.js:13]: This function must return token, not rewrite header

File: passport.js [hunk 15]
```diff
@@ -0,0 +1,61 @@
+const GoogleStrategy = require('passport-google-oauth').OAuth2Strategy;
+const LinkedInStrategy = require('passport-linkedin-oauth2').Strategy;
+
+module.exports = passport => {
```
Comment (slavasgod) [inline passport.js]: Migrate this file to config folder, because it is passport config.

File: passport.js [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    done(null, user);
+  });
+  passport.use(
+    new GoogleStrategy(
+      {
+        clientID: process.env.GOOGLE_CLIENT_ID,
+        clientSecret: process.env.GOOGLE_SECRET,
+        callbackURL: process.env.SERVER_HOST + 'api/users/auth/google/callback',
+      },
+      (accessToken, refreshToken, profile, done) => {
```
Comment (slavasgod) [inline passport.js]: Split this callbacks, cause you use it in 2 places

**Review by slavasgod — CHANGES_REQUESTED** [review]

File: api/middlewares/refreshAccessToken.js [hunk 17]
```diff
@@ -1,16 +0,0 @@
-const jwtDecode = require('jwt-decode');
```
Comment (melonholy) [inline api/middlewares/refreshAccessToken.js:1]: return this file

File: api/middlewares/refreshAccessToken.js [hunk 18]
```diff
@@ -1,16 +0,0 @@
...
-const jwt = require('jsonwebtoken');
```
Comment (melonholy) [inline api/middlewares/refreshAccessToken.js:2]: leave this lib, delete jwt decode

File: config/passport.js [hunk 19]
```diff
@@ -1,6 +1,22 @@
...
+    id,
+    name: { givenName, familyName },
+    emails,
+  } = profile;
+  const userData = {
+    id: id,
+    email: emails[0].value,
+    firstName: givenName,
+    lastName: familyName,
+    token: accessToken,
```
Comment (melonholy) [inline config/passport.js]: delete this field

File: config/passport.js [hunk 20]
```diff
@@ -1,6 +1,22 @@
...
+  };
```
Comment (melonholy) [inline config/passport.js:15]: rewrite db structure to apply external users and make password field non-nullable

**Review by slavasgod — APPROVED** [review]

**Review by OlegPanasyuk — APPROVED** [review]

**Pull request merged**